src/dynamichmc_inference.jl

```diff
@@ -25,16 +27,28 @@ Base.@kwdef struct DynamicHMCPosterior{TA,TP,TD,TT,TR,TS,TK}
     σ_priors::TS
     "Keyword arguments passed on the the ODE solver `solve`."
     solve_kwargs::TK
+    obsvbls::TO
+    sample_u0::Bool
+    adkind::AD
 end
 
 function (P::DynamicHMCPosterior)(θ)
     @unpack parameters, σ = θ
-    @unpack algorithm, problem, data, t, parameter_priors, σ_priors, solve_kwargs = P
-    prob = remake(problem, u0 = convert.(eltype(parameters), problem.u0), p = parameters)
-    solution = solve(prob, algorithm; solve_kwargs...)
-    any((s.retcode ≠ :Success && s.retcode ≠ :Terminated) for s in solution) && return -Inf
-    log_likelihood = sum(sum(logpdf.(Normal.(0.0, σ), solution(t) .- data[:, i]))
-                         for (i, t) in enumerate(t))
+    @unpack algorithm, problem, data, t, parameter_priors = P
+    @unpack σ_priors, solve_kwargs, obsvbls, sample_u0, adkind = P
+    nu, T = length(problem.u0), eltype(parameters)
+    u0 = convert.(T, sample_u0 ? parameters[1:nu] : problem.u0)
+    p = convert.(T, sample_u0 ? parameters[(nu + 1):end] : parameters)
+    _saveat = t === nothing ? Float64[] : t
+    prob = remake(problem, u0 = Tracker.data.(u0), p = Tracker.data.(p))
+    solution = solve(prob, algorithm; saveat = _saveat, solve_kwargs...)
+    if T <: Tracker.TrackedReal || adkind == Val(:Zygote)
+        solution′ = concrete_solve(problem, algorithm, u0, p; saveat = _saveat, solve_kwargs...)
+    else
+        solution′ = solution
+    end
+    any((s.retcode ≠ :Success && s.retcode ≠ :Terminated) for s in solution) && return T(0) * sum(σ) + T(-1e10)
```

Comment:
why `-1e10`?

Reply:
Because some optimization packages such as `Optim.jl` won't accept `Inf` values ( `Optim` is used in `DynamicHMC.jl`'s warmup phases). `T(0) * sum(σ)` is used because `Tracker` will error if the output is constant and `Zygote` will give a `nothing` for its gradient which is incompatible with the entire pipeline.

Optimization can be used for MAP too. 
```
function get_nlogp(model)
    vi = Turing.VarInfo(model)
    function nlogp(θ)
        spl = Turing.SampleFromPrior()
        vi′ = Turing.VarInfo(vi, spl, θ)
        model(vi′, spl)
        -Turing.getlogp(vi′)
    end
    back(θ) = l̄ -> (-Turing.gradient_logp_reverse(θ, vi, model)[2],)
    @eval Zygote.@adjoint $nlogp(θ) = $nlogp(θ), $back(θ)
    return nlogp
end
res_turing = turing_inference(...)
)
nlogp = get_nlogp(res_turing.info.model)
DiffEqFlux.sciml_train(nlogp, ...)
```

Reply:
Optim accepts and works with Infs just fine by rejecting those steps. Using a largeish value might be the cause of the test failures?